main.rb

```diff
@@ -0,0 +1,142 @@
+
+module Enumerable
+  def my_each
+    return to_enum :my_each unless block_given?
+
+    index = 0
+    while index < size
+      if is_a? Array
+        yield self[index]
+      elsif is_a? Hash
+        yield keys[index], self[keys[index]]
+      elsif is_a? Range
+        yield to_a[index]
+      end
+      index += 1
+    end
+  end
+
+  def my_each_with_index
+    return to_enum :my_each unless block_given?
+
+    index = 0
+    while index < size
+      if is_a? Array
+        yield self[index], index
+      elsif is_a? Hash
+        yield keys[index], self[keys[index]]
+      elsif is_a? Range
+        yield to_a[index], index
+      end
+      index += 1
+    end
+  end
+
+  def my_select
+    return to_enum :my_select unless block_given?
+
+    if is_a? Array
+      results = []
+      my_each { |x| results << x if yield x }
+    else
+      results = {}
+      my_each { |k, v| results[k] = v if yield k, v }
+    end
+    results
+  end
+
+  def my_any?(args = nil)
+    block_is_true = false
+    if block_given?
+      my_each { |x| block_is_true = true if yield(x) }
+    elsif args.is_a? Regexp
+      my_each { |x| block_is_true = true if x.match(args) }
+    elsif args.is_a? Module
+      my_each { |x| block_is_true = true if x.is_a?(args) }
+    elsif !block_given? || args.nil?
+      my_each { |x| block_is_true = true unless x.nil? || x == false }
+    else
+      my_each { |x| block_is_true = true if x.nil? || x == false }
+    end
+    block_is_true
+  end
+
+  def my_map(&proc)
+    return to_enum :my_map unless block_given?
+
+    results = []
+    my_each { |x| results << proc.call(x) }
+    results
+  end
+
+  def my_all?(args = nil)
+    block_is_true = true
+    if block_given?
+      my_each { |x| block_is_true = false unless yield(x) }
+    elsif args.is_a? Regexp
+      my_each { |x| block_is_true = false unless x.match(args) }
+    elsif args.is_a? Module
+      my_each { |x| block_is_true = false unless x.is_a?(args) }
+    elsif !block_given? || args.nil?
+      my_each { |x| block_is_true = false unless x.nil? || x == false }
+    else
+      my_each { |x| block_is_true = false unless x == true }
+    end
+    block_is_true
+  end
+
+  def my_count(args = nil)
+    counter = 0
+    if block_given?
+      my_each { |x| counter += 1 if yield(x) }
+    elsif !args.nil?
+      my_each { |x| counter += 1 if args == x }
+    else
+      counter = length
+    end
+    counter
+  end
+
+  def my_none?(args = nil)
+    block_is_true = true
+    if block_given?
+      my_each { |x| block_is_true = false if yield(x) }
+    elsif args.is_a? Regexp
+      my_each { |x| block_is_true = false if x.match(args) }
+    elsif args.is_a? Module
+      my_each { |x| block_is_true = false if x.is_a?(args) }
+    elsif !block_given? || args.nil?
+      my_each { |x| block_is_true = false unless x.nil? || x == false }
+    else
+      my_each { |x| block_is_true = false unless x == false }
+    end
+    block_is_true
+  end
```

Comment:
* [ ] `my_none?` is identical and returns the same thing as ruby's [`none?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-none-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_none.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

* [ ] my_none? when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern
  ```ruby
    array = %w[dog door rod blade]
    p array.my_none?(5) == array.none?(5) # false
  ```